kappa.js

```diff
@@ -0,0 +1,319 @@
+    if (this._source.filter) this._transform.push(this._source.filter.bind(this._source))
+    if (this._source.transform) this._transform.push(this._source.transform.bind(this._source))
+    if (this.opts.filter) this._transform.push(this.opts.filter)
+    if (this.opts.transform) this._transform.push(this.opts.transform)
+    if (this._view.filter) this._transform.push(this._view.filter.bind(this._view))
+    if (this._view.transform) this._transform.push(this._view.transform.bind(this._view))
+
+    this._opened = false
+    this.open = thunky(this._open.bind(this))
+  }
+
+  get version () {
+    return this._view.version || 1
+  }
+
+  _open (cb = noop) {
+    if (this._opened) return cb()
+    const self = this
+
+    let pending = 1
+    if (this._view.open) ++pending && this._view.open(this, onopen)
+    if (this._source.open) ++pending && this._source.open(this, onopen)
+    onopen()
+
+    function onopen () {
+      if (--pending !== 0) return
+      if (self._source.fetchVersion) {
+        self._source.fetchVersion((err, version) => {
+          if (err) return ondone()
+          if (!version) return self._source.storeVersion(self.version, ondone)
+          if (version !== self.version) {
+            self.reset(() => self._source.storeVersion(self.version, ondone))
+          } else ondone()
+        })
+      } else ondone()
+    }
+
+    function ondone () {
+      self._opened = true
+      self._run()
+      cb()
+    }
+  }
+
+  close (cb) {
+    const self = this
+    this.pause()
+    this._closing = true
+    if (this.status === Status.Running) return this.once('ready', close)
+    else close()
+    function close () {
+      let pending = 1
+      if (self._source.close) ++pending && self._source.close(cb)
+      if (self._view.close) ++pending && self._view.close(cb)
+      done()
+      function done () {
+        if (--pending !== 0) return
+        self._closing = false
+        self._opened = false
+        cb()
+      }
+    }
+  }
+
+  ready (cb, waitForSource) {
+    const self = this
+    if (!this._opened) return this.open(() => this.ready(cb))
+
+    setImmediate(() => {
+      if (this.source.ready) this.source.ready(onsourceready)
+      else if (this._source.ready) this._source.ready(onsourceready)
+      else onsourceready()
+    })
+
+    function onsourceready () {
+      process.nextTick(() => {
+        if (self.status === Status.Ready) process.nextTick(cb)
+        else self.once('ready', cb)
+      })
+    }
+  }
+
+  pause () {
+    this.status = Status.Paused
+  }
+
+  resume () {
+    if (this.status !== Status.Paused) return
+    this.status = Status.Ready
+    if (!this._opened) return this.open()
+    this._run()
+  }
+
+  reset (cb = noop) {
+    const self = this
+    this.pause()
+    let pending = 1
+    process.nextTick(() => {
+      if (this._view.clearIndex) ++pending && this._view.clearIndex(done)
+      if (this._source.reset) ++pending && this._source.reset(done)
```

Comment:
Yup I think it should be required. Currently this is not enforced, but this would mean that after increasing the view's version number, if the source does not implement `reset`, things go wrong (the view would continue receiving new messages instead of starting from the beginning). I think there might be usecases for "non-historic" sources, but that has untested implications. I marked it in the documentation as required. Likely we want to assert this in the code as well?

Reply:
What do you think about checking sources/flows/views on `kappa.use()` to ensure they have all of the properties set, and fail if not? I bet that will catch a lot of mistakes.